### DISCUSSION TITLE
Modal: Add onClose() callback method

### DISCUSSION THREAD

**Discussion by scottanthonymurray** [indiana-university/rivet-source#415] [Feature and enhancement requests]
There are a couple of problems with Modals that are difficult to solve with the current implementation:

1. How do you know when the Modal closes (via clicking backdrop or keying escape or clicking a close button)?
2. How do you prevent a Model from closing?

Why it’s important to know (in the case of AdRx Reminders):

1. Modals are not created and destroyed upon opening and closing. They’re reused, which should hopefully improve performance a little. Upon close, the Modal is cleaned up and restored, ready for the next use. Given the Rivet CSS focus animations, this cleanup work cannot be done upon open without potentially having a flash of incorrectly styled content.
2. When the Modal submits, we suspend the user in place until the server responds (see below for screenshots). We don’t want the user to close the Modal during this pending state given the request could fail; if that would happen, then the user would not see that the fail happened. This ensures we know upfront if it was successful or not. If it succeeds, the Modal closes. If it fails, the Modal stays open and focus moves to an error alert at the top of the Modal.

How I’ve solved these problems:

1. Use MutationObserver to listen for `aria-hidden` updates and infer when the Modal closes.
2. Disable all buttons. Stretch an empty `<div>` across the top of the viewport to block Modal backdrop clicks (and consequently all clicks altogether).

I outline both of these problems in this same issue because I think there’s a common solution. (Feel free to break it apart, if needed.) Add an `onClose` method that receives a callback. The callback could provide a custom close event or event-like object, which could allow the close to be cancelled. For example:

```js
// API
Modal.onClose('modalId', callback)

// Example 1
Modal.onClose('reminder-modal', () => cleanUp())

// Example 2
Modal.onClose('reminder-modal', (event) => {
  if (isPending) {
    event.preventDefault()
  }
})
```

***

New Reminder modal ready to be submitted.

![screen shot 2018-09-04 at 10 38 38 am](https://media.github.iu.edu/user/3998/files/a51d0f62-b03b-11e8-8af1-f7b7bc99293a)

New Reminder modal waiting on server response after submission.

![screen shot 2018-09-04 at 10 38 45 am](https://media.github.iu.edu/user/3998/files/b9efb124-b03b-11e8-8c9c-4a1afc6da356)

**Comment by zttodd** [comment]
This should be included in a future release when re-evaluating the modal component.
